src/sentry/migrations/0378_backfill_perf_issue_events_issue_platform.py

```diff
@@ -0,0 +1,648 @@
+
+    def __enter__(self) -> "FileBackedProjectProgress":
+        return self
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        try:
+            for progress in self.values():
+                progress.close()
+        except ValueError:
+            pass
+
+    def get_or_init(self, group_id: int) -> ProjectGroupTotalProgress:
+        if self.get(group_id):
+            return self.__getitem__(group_id)
+        else:
+            progress = ProjectGroupTotalProgress(self.project_id, group_id)
+            self.__setitem__(group_id, progress)
+            return progress
+
+
+def backfill_by_project(project_id: int, Group: Any, GroupHash: Any, dry_run: bool) -> None:
+    next_offset: Optional[int] = 0
+
+    # TODO: need to revisit this since we'll be inserting data to search_issues dataset as we backfill
+    #        gonna need to figure out how to ensure we don't process dupes
+    already_processed_events: Mapping[int, typing.Set[str]] = defaultdict(
+        set
+    )  # group_id: Set[event_id]
+
+    # it's probably ok to query all the events here instead of paginating since we should be running this script
+    # closely after we cut-over
+    # for issue_platform_events in _query_issue_platform_events(
+    #     project_id, start=ISSUE_PLATFORM_INGEST_START_DATETIME, end=END_DATETIME
+    # ):
+    #     already_processed_events[issue_platform_events["group_id"]].add(
+    #         issue_platform_events["event_id"]
+    #     )
+
+    with FileBackedProjectProgress(project_id) as previous_progress:
+        while next_offset is not None:
+            rows, next_offset = _query_performance_issue_events(
+                project_id=project_id,
+                start=START_DATETIME,
+                end=END_DATETIME,
+                offset=next_offset,
+            )
+
+            for row in rows:
+                group_id = row["group_id"]
+                event_id = row["event_id"]
+                finish_ts = row["finish_ts"]
+
+                # avoid re-ingesting the same transaction since we'll be dual-writing
+                if event_id not in already_processed_events[group_id]:
+                    group_progress: ProjectGroupTotalProgress = previous_progress.get_or_init(
+                        group_id
+                    )
+                    error_maybe = group_progress.consume_error()
+                    if error_maybe is None:
+                        already_processed = group_progress.already_processed(finish_ts, event_id)
+                        if already_processed:
+                            continue
+                        else:
+                            backfill_perf_issue_occurrence(
+                                row,
+                                Group,
+                                GroupHash,
+                                print_success
+                                if not WRITE_TO_FILE
+                                else group_progress.save_success_to_file,
```

Comment:
Do we actually need to track progress by group? I would have thought it'd be enough to track the last timestamp/event in the project.

Reply:
Yeah, this was an artifact of how I initially implemented it. With the new style, we don't need to track progress by group_id. I think its still worthwhile to write it out to a file since its not hurting anything, but we won't be reading from it at all.